File: backend/src/main/java/com/flock/journal/homework/HomeworkService.java
package com.flock.journal.homework;

import java.util.List;
import java.util.Optional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public class HomeworkService {

  private final HomeworkRepository homeworkRepository;

  @Autowired
  public HomeworkService(HomeworkRepository homeworkRepository) {
    this.homeworkRepository = homeworkRepository;
  }

  public List<Homework> getAllHomeworks() {
    return homeworkRepository.findAll();
  }

  public List<Homework> getHomeworksByGroupDisciplineProfessor(
      Long groupId,
      Long disciplineId,
      Long professorId
  ) {
    return homeworkRepository.findByGroupDisciplineProfessor(
        groupId,
        disciplineId,
        professorId
    );
  }

  public List<Homework> getHomeworkByDisciplineAndGroupAndStudent(
      Long disciplineId,
      Long groupId,
      Long studentId
  ) {
    return homeworkRepository.findByDisciplineAndGroupAndStudent(
        disciplineId,
        groupId,
        studentId
    );
  }

  public Optional<Homework> getHomeworkById(Long id) {
    return homeworkRepository.findById(id);
  }

  public Homework saveHomework(Homework homework) {
    return homeworkRepository.save(homework);
  }

  public Optional<Homework> updateHomework(Long id, Homework updatedHomework) {
    Optional<Homework> homework = getHomeworkById(id);
    if (homework.isPresent()) {
      Homework existingHomework = homework.get();
      if (updatedHomework.getLesson() != null) {
        existingHomework.setLesson(updatedHomework.getLesson());
      }
      if (updatedHomework.getProfessor() != null) {
        existingHomework.setProfessor(updatedHomework.getProfessor());
      }
      existingHomework.setDescription(updatedHomework.getDescription());

      return Optional.of(homeworkRepository.save(existingHomework));
    } else {
      return Optional.empty();
    }
  }
  
  public void deleteHomework(Long id) {
    homeworkRepository.deleteById(id);
  }
}

File: frontend/src/services/homeworkService.js
import api from "./api";
import {createAuthHeaders} from "./authService";

export const getAllHomeworks = async () => {
    try {
        const headers = createAuthHeaders();
        const response = await api.get("/homeworks", {headers});

        return response.data;
    } catch (error) {
        throw new Error(error.response.data.error);
    }
};

export const getHomeworkById = async (id) => {
    try {
        const headers = createAuthHeaders();
        const response = await api.get(`/homeworks/${id}`, {headers});

        return response.data;
    } catch (error) {
        throw new Error(error.response.data.error);
    }
};

export const getHomeworksByGroupDisciplineProfessor = async (groupId, disciplineId, professorId) => {
    try {
        const headers = createAuthHeaders();
        const response = await api.get(`/homeworks/groups/${groupId}/${disciplineId}/${professorId}`, {headers});

        return response.data;
    } catch (error) {
        throw new Error(error.response.data.error);
    }
};

export const getHomeworksByDisciplineStudentGroup = async (disciplineId, groupId, studentId) => {
    try {
        const headers = createAuthHeaders();
        const response = await api.get(`/homeworks/${disciplineId}/${groupId}/${studentId}`, {headers});

        return response.data;
    } catch (error) {
        throw new Error(error.response.data.error);
    }
};

export const createHomework = async (homework) => {
    try {
        const headers = createAuthHeaders();
        const response = await api.post("/homeworks", homework, {headers});

        return response.data;
    } catch (error) {
        throw new Error(error.response.data.error);
    }
};

export const updateHomework = async (id, updatedHomework) => {
    try {
        const headers = createAuthHeaders();
        const response = await api.put(`/homeworks/${id}`, updatedHomework, {headers});

        return response.data;
    } catch (error) {
        throw new Error(error.response.data.error);
    }
};

export const deleteHomework = async (id) => {
    try {
        const headers = createAuthHeaders();
        await api.delete(`/homeworks/${id}`, {headers});
    } catch (error) {
        throw new Error(error.response.data.error);
    }
};
